gaia-fast-list.js

```diff
@@ -815,12 +1060,132 @@ Internal.prototype = {
+
+  scheduler.attachDirect(this.el, touchstart, this.onTouchStart);
+  this.el.addEventListener('click', this.onClick, true);
+
+  this.render();
+}
+
+Picker.prototype = {
+  render() {
+    var letters = 'abcdefghijklmnopqrstuvwxyz#';
+    var length = letters.length;
+
+    for (var i = 0; i < length; i++) {
+      var letter = letters[i];
+      var el = document.createElement('a');
+      el.textContent = letters[i];
+      el.href = `#gfl-section-${letter}`;
+      this.el.appendChild(el);
+      this.els.items.push(el);
+    }
+  },
+
+  addEventListener(name, fn) { this.el.addEventListener(name, fn); },
+  removeEventListener(name, fn) { this.el.removeEventListener(name, fn); },
+
+  onTouchStart(e) {
+    debug('touch start');
+    this.height = this.el.clientHeight;
+    this.els.allItems = this.getAllItems();
+    this.itemHeight = this.height / this.els.allItems.length;
+    this.offset = this.el.getBoundingClientRect().top;
+
+    this.update(e);
+    scheduler.attachDirect(window, touchmove, this.onTouchMove);
+    scheduler.attachDirect(window, touchend, this.onTouchEnd);
+    this.emit('started');
+  },
+
+  onTouchMove(e) {
+    debug('touch move');
+    var fast = (e.timeStamp - this.lastUpdate) < 50;
+    if (!fast) this.update(e);
+  },
+
+  onTouchEnd(e) {
+    debug('touch end');
+    this.update(e);
+    scheduler.detachDirect(window, touchmove, this.onTouchMove);
+    scheduler.detachDirect(window, touchend, this.onTouchEnd);
+    this.emit('ended');
+  },
+
+  onClick(e) {
+    e.preventDefault();
+    e.stopPropagation();
+  },
+
+  update(e) {
+    debug('update', this.selectedIndex);
+    var allItems = this.els.allItems;
+    var pageY = e.pageY || e.changedTouches[0].pageY;
+    var y = pageY - this.offset;
+    var index = Math.floor(y / this.itemHeight);
+
+    // clamp within index range
+    index = Math.max(0, Math.min(allItems.length - 1, index));
+
+    // abort if new index is same as current
+    if (index === this.selectedIndex) return;
+
+    this.selectedIndex = index;
+    this.selected = allItems[index];
+    this.lastUpdate = e.timeStamp;
+    this.emit('picked');
+  },
+
+  getAllItems() {
+    var light = [].slice.call(this.els.content.getDistributedNodes());
+    var shadow = this.els.items;
+    return light.concat(shadow);
+  },
+
+  emit(name) {
+    this.el.dispatchEvent(new CustomEvent(name, { bubbles: false }));
+  },
+
+  destroy() {
+    this.els.items.forEach(el => el.remove());
+    scheduler.detachDirect(this.el, touchstart, this.onTouchStart);
+    this.el.removeEventListener('click', this.onClick, true);
+  },
+};
+
 /**
  * Exports
  */
 
 module.exports = component.register('gaia-fast-list', GaiaFastListProto);
 
+/**
+ * Utils
+ */
+
+function throttle(callback, limit) {
+  var wait = false;
+  return function() {
+    if (!wait) {
+      callback.apply(this, arguments);
+      setTimeout(() => wait = false, limit);
+      wait = true;
+    }
+  };
+}
+
```

Comment:
as said on IRC we'll probably remove this but FTR the issue is that we can "loose" the last call where ideally we want to skip intermediate calls but honor the last one.